textsearch/python/csrc/suffix_array.cc

```diff
@@ -6,42 +6,49 @@
 #include "pybind11/numpy.h"
 #include "pybind11/stl.h"
 #include "textsearch/csrc/suffix_array.h"
+#include <iostream>
 #include <limits>
 
 namespace fasttextsearch {
 
 template <typename T>
-py::array_t<T> PybindSuffixArrayHelper(py::array_t<T> &input) {
+static py::array_t<int64_t> PybindSuffixArrayHelper(py::array_t<T> &input) {
 
   py::buffer_info input_buf = input.request();
 
   if (input_buf.ndim != 1)
     throw std::runtime_error("Input MUST be a one dimension array");
 
-  T input_length = input_buf.size;
-  T seq_len = input_length - 3;
+  int64_t input_length = input_buf.size;
+  int64_t seq_len = input_length - 3;
+  T *input_data = static_cast<T *>(input_buf.ptr);
 
-  auto suffix_array = py::array_t<T>(seq_len);
+  // copy input to int64_t array
+  auto input64 = py::array_t<int64_t>(input_length);
+  py::buffer_info input64_buf = input64.request();
+  int64_t *input64_data = static_cast<int64_t *>(input64_buf.ptr);
+  for (int64_t i = 0; i < input_length; ++i)
+    input64_data[i] = input_data[i];
 
```

Comment:
> Actually, on the other hand, the memory for the symbols will not dominate, and the time for the suffix-array part will also not dominate, so at the Python level, for now we could just copy them to the index type, e.g. int64. So no need to mess with teh C++ code I think.

For above comment, do you mean copy input ndarray to int64_t and just use int64_t in our internal C++ code. I have not found the interface like `to()` in `py::array_t` so just copy the elements in a for loop.

Reply:
Can we use a wrapper in Python, which converts the input to `np.int64` before passing to C++?

Reply:
> Can we use a wrapper in Python, which converts the input to `np.int64` before passing to C++?

Of course we can do it in Python, but I think it make no difference as my current code, do you think there is a more efficient implementation to copy np.int16 array to np.int64 in numpy?
If we do it in Python, we have to pass max-symbol to C++ too, as `max-symbol` is derived from dtype (i.e. np.iinfo(dtype).max).

Reply:
It will reduce the amount of C++ code.

Reply:
Less code -> fewer errors

---

In your current c++ code, you are assuming the input array is contiguous in memory, which is not always the case, I think.

Reply:
> It will reduce the amount of C++ code.

Ok, if so, wo can also remove the template from pybind code, I will change it. Thanks!